core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperator.java

```
@@ -0,0 +1,218 @@
+   * If rareTopFlag is false, result for rare will be returned.
+   */
+  @Getter
+  private final Boolean rareTopFlag;
+  @Getter
+  private final Integer noOfResults;
+  @Getter
+  private final List<Expression> fieldExprList;
+  @Getter
+  private final List<Expression> groupByExprList;
+
+  @EqualsAndHashCode.Exclude
+  private final Group group;
+  @EqualsAndHashCode.Exclude
+  private Iterator<ExprValue> iterator;
+
+  private static final Integer DEFAULT_NO_OF_RESULTS = 10;
+
+
+  public RareTopNOperator(PhysicalPlan input, Boolean rareTopFlag, List<Expression> fieldExprList,
+      List<Expression> groupByExprList) {
+    this(input, rareTopFlag, DEFAULT_NO_OF_RESULTS, fieldExprList, groupByExprList);
+  }
+
+  /**
+   * RareTopNOperator Constructor.
+   *
+   * @param input           Input {@link PhysicalPlan}
+   * @param rareTopFlag     Flag for Rare/TopN command.
+   * @param noOfResults     Number of results
+   * @param fieldExprList   List of {@link Expression}
+   * @param groupByExprList List of group by {@link Expression}
+   */
+  public RareTopNOperator(PhysicalPlan input, Boolean rareTopFlag, int noOfResults,
+      List<Expression> fieldExprList,
+      List<Expression> groupByExprList) {
+    this.input = input;
+    this.rareTopFlag = rareTopFlag;
+    this.noOfResults = noOfResults;
+    this.fieldExprList = fieldExprList;
+    this.groupByExprList = groupByExprList;
+    this.group = new Group();
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitRareTopN(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return iterator.hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    return iterator.next();
+  }
+
+  @Override
+  public void open() {
+    super.open();
+    while (input.hasNext()) {
+      group.push(input.next());
+    }
```

Should we do something to indicate that input is no longer usable or throw an error if open() is called more than once?

---

I am not sure if it is possible to call open() more than once for a single request. AggregationOperator & SortOperator also override open() and it doesn't throw an error so I assumed it's not required.

---

I would suggest using Precondition.checkNull(iterator) here to fail fast if open() accidentally gets called more than once.

---

open() gets called only once